URL scheme handler is not used when redirecting from http(s) to the custom scheme

cdumez opened the pull request:
#### c839c8024f7775cc5326bd8a5b10639a40620111
<pre>
URL scheme handler is not used when redirecting from http(s) to the custom scheme
<a href="https://bugs.webkit.org/show_bug.cgi?id=266751">https://bugs.webkit.org/show_bug.cgi?id=266751</a>
<a href="https://rdar.apple.com/117891282">rdar://117891282</a>

Reviewed by John Wilander and Alex Christensen.

We updated our process-swapping logic in a recent iOS update and it broke the flow
to add a new payment method in some cases. The flow navigates from a https:// URL
to a another scheme which the client app is supposed to handle via a URL scheme
handler. In older iOS version, we used to process-swap for such navigations and
the navigation in the new WebProcess would correctly use the URL scheme handler.

However, in newer iOS versions, we sometimes don&apos;t process-swap. This is the case
for example when the process has not committed a provisional load yet. Because
we don&apos;t process-swap, the redirect happens in the old process, the URL scheme
handler is not used and we ask the network to try and follow the redirect, which
fails.

To address the issue, I am adding a check in processForNavigationInternal() to
make sure we process-swap when redirecting to a new scheme and this scheme has a
custom handler registered.

This is covered by a new API test.

* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processForNavigationInternal):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKURLSchemeHandler-1.mm:

Canonical link: <a href="https://commits.webkit.org/272427@main">https://commits.webkit.org/272427@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2ce51f92b10837617a92e7446e1fa59b563711a4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/31584 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/10262 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/33298 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/34076 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/28608 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/32361 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/12616 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/7513 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/28216 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/31931 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/8640 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/28192 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/7450 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/7610 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/28099 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/35421 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/28705 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/28549 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/33738 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/7699 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/5713 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/31589 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/9354 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7417 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8384 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/8205 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

johnwilander commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -1990,6 +1990,12 @@ std::tuple<Ref<WebProcessProxy>, SuspendedPageProxy*, ASCIILiteral> WebProcessPo
+    // process swap, so that we end up with a fresh navigation to the custom scheme and so that the
+    // custom protocol handler is used (rdar://117891282).
+    if (navigation.currentRequestIsRedirect() && navigation.originalRequest().url().protocol() != targetURL.protocol() && page.urlSchemeHandlerForScheme(targetURL.protocol().toString()))
+        return { createNewProcess(), nullptr, "Redirect to a custom protocol"_s };
```
We don't know that it's a *custom* protocol here, right? Just that it's a different protocol. We should change the string to reflect that.

cdumez replied:
We called page.urlSchemeHandlerForScheme() and it returned non-null, so it is a custom protocol that the app registered a handler for.

cdumez replied:
It is not just "any" different protocol.

johnwilander replied:
I don't see where the dependency on custom protocol comes in.  Is it `page.urlSchemeHandlerForScheme()` that scopes this to only custom protocols? That sounds implicit and I would change the string to match the naming of that function more closely.

cdumez replied:
Maybe I'm using the wrong term? I used "custom protocol" to mean a "scheme for which the client registered their own handler". Maybe I can change the logging to use that instead, because that's exactly what I'm checking for.

johnwilander replied:
Yeah, that sounds much better.

johnwilander reviewed (approved):
Approved pending my comments.

johnwilander commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKURLSchemeHandler-1.mm:
```diff
@@ -1731,3 +1731,24 @@ - (void)userContentController:(WKUserContentController *)userContentController d
+    [webView loadRequest:server.request("/redirect.html"_s)];
+
+    TestWebKitAPI::Util::run(&schemeHandledCalled);
+}
```
It lost my comment here. I was asking for 1) a second test that goes the opposite direction, i.e. custom protocol to HTTP(S), and 2) a test case with another 3xx redirect such as 301.

cdumez replied:
Redirection from custom scheme to HTTP(S) is something I'm trying to support here. It's certainly not what broke and I have no idea if this used to work or not.

johnwilander replied:
Better to know than not. I'm not asking for further behavior changes, just testing that we don't regress something else.

cdumez replied:
My change can't change the behavior of redirections to HTTP(s) since it is gated on having a custom scheme handler for the destination URL scheme.

cdumez replied:
I haven't been able to write such a test yet. I asked Brady if it's possible or not. As far as I can tell, when the custom scheme handler triggers a redirect, it has to provide the response for the redirected URL as well. So we can't really redirect from custom scheme to HTTP it seems.

cdumez replied:
I just got confirmation from Brady that redirecting from custom scheme to HTTP is just not possible with the current API.

> You can re-write the URL with a redirect, but then need to keep handling the task yourself.

So the scheme handler can trigger a server redirect but then it has to handle that redirect itself, it cannot just say "redirect to HTTP and stop using me" for this navigation.

johnwilander reviewed (approved):
See my comment on the test case too. GitHub defaults to "Start a review" which is why a new review was opened.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKURLSchemeHandler-1.mm:
```diff
@@ -1731,3 +1731,34 @@ - (void)userContentController:(WKUserContentController *)userContentController d
+    runRedirectToHandledSchemeTest(301);
+}
+
+TEST(URLSchemeHandler, Redirect302FromHTTPToHandledScheme)
```
These could both be in the same test.

cdumez replied:
But this makes it clearer if it fails, which one actually failed.

cdumez replied:
Otherwise, you'll get a failure in the static function and you won't know which redirect it was for.

cdumez committed:
URL scheme handler is not used when redirecting from http(s) to the custom scheme
https://bugs.webkit.org/show_bug.cgi?id=266751
rdar://117891282

Reviewed by John Wilander and Alex Christensen.

We updated our process-swapping logic in a recent iOS update and it broke the flow
to add a new payment method in some cases. The flow navigates from a https:// URL
to a another scheme which the client app is supposed to handle via a URL scheme
handler. In older iOS version, we used to process-swap for such navigations and
the navigation in the new WebProcess would correctly use the URL scheme handler.

However, in newer iOS versions, we sometimes don't process-swap. This is the case
for example when the process has not committed a provisional load yet. Because
we don't process-swap, the redirect happens in the old process, the URL scheme
handler is not used and we ask the network to try and follow the redirect, which
fails.

To address the issue, I am adding a check in processForNavigationInternal() to
make sure we process-swap when redirecting to a new scheme and this scheme has a
custom handler registered.

This is covered by a new API test.

* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processForNavigationInternal):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKURLSchemeHandler-1.mm:

Canonical link: https://commits.webkit.org/272427@main
